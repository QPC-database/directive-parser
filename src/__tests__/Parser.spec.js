const Parser = require('../Parser');

test('Parses directive', () => {
    const { ast } = new Parser('@RootComponent').parse();
    const [node] = ast.body;

    expect(node.type).toEqual('annotation');
    expect(node.value).toEqual('RootComponent');
});

test('Parses assignment with list as rhs', () => {
    const { ast } = new Parser('foo = bizz, bazz, buzz').parse();
    const [node] = ast.body;

    expect(node.type).toEqual('assignment');

    const list = node.rhs;
    expect(list.type).toEqual('list');
    ['bizz', 'bazz', 'buzz'].forEach((ident, i) => {
        expect(list.items[i].value).toEqual(ident);
    });
});

test('Parses assignment with identifier as rhs', () => {
    const { ast } = new Parser('foo = bizz').parse();
    const [node] = ast.body;

    expect(node.type).toEqual('assignment');
    expect(node.rhs.type).toEqual('identifier');
    expect(node.rhs.value).toEqual('bizz');
});

test('Parses assignment with string as rhs', () => {
    const str = 'str!ng with rand0m chars ðŸ˜Ž';
    const { ast } = new Parser(`foo = "${str}"`).parse();
    const [node] = ast.body;

    expect(node.type).toEqual('assignment');
    expect(node.rhs.type).toEqual('string');
    expect(node.rhs.value).toEqual(str);
});

test('Parses comments with leading asterisks', () => {
    const { ast } = new Parser(`
        /**
         * @RootComponent
         */
    `).parse();
    const [node] = ast.body;

    expect(node.type).toEqual('annotation');
    expect(node.value).toEqual('RootComponent');
});

test('Parses the kitchen sink', () => {
    const { ast } = new Parser(`
        /**
         * @RootComponent
         * foo = bizz, bazz, buzz
         * bar = car
         * jar = "test"
         */
    `).parse();
    expect(ast).toMatchSnapshot();
});

test('Fails when directive is unrecognized type', () => {
    const { error } = new Parser('@Unknown').parse();
    expect(error.message).toEqual('Unrecognized Directive: Unknown');
});

test('Fails when string is not terminated', () => {
    const { error } = new Parser('foo = "bar').parse();
    expect(error.message).toBe('Unterminated string encountered');
});

test('Fails when list has dangling comma that creates ambiguity with ident on next line', () => {
    const { error } = new Parser(`
        @RootComponent
        pageTypes = foo, bizz,
        description = "hey"
    `).parse();
    expect(error.message).toEqual(
        'Encountered illegal assignment in an unterminated list'
    );
});

test('Parses when dangling comma in list does not create ambiguity because EOF', () => {
    const { ast } = new Parser(`
        @RootComponent
        pageTypes = foo, bizz,
    `).parse();
    const [, node] = ast.body;
    expect(node.type).toEqual('assignment');
    expect(node.rhs.items.length).toEqual(2);
});

test('Parses multiple pieces of metadata on a single line', () => {
    const { ast: { body }, error } = new Parser(`
        @RootComponent pageTypes = foo, bizz description = "test"
    `).parse();

    const [first, second, third] = body;
    expect(first.type).toEqual('annotation');
    expect(first.value).toEqual('RootComponent');

    expect(second.type).toEqual('assignment');
    expect(second.rhs.items.length).toEqual(2);

    expect(third.type).toEqual('assignment');
    expect(third.rhs.value).toEqual('test');
});
