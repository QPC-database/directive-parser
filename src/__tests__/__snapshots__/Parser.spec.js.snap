// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parses the kitchen sink 1`] = `
Object {
  "body": Array [
    Object {
      "type": "annotation",
      "value": "RootComponent",
    },
    Object {
      "lhs": Object {
        "location": Object {
          "end": Object {
            "column": 14,
            "line": 4,
          },
          "start": Object {
            "column": 11,
            "line": 4,
          },
        },
        "type": "identifier",
        "value": "foo",
      },
      "rhs": Object {
        "items": Array [
          Object {
            "location": Object {
              "end": Object {
                "column": 21,
                "line": 4,
              },
              "start": Object {
                "column": 17,
                "line": 4,
              },
            },
            "type": "identifier",
            "value": "bizz",
          },
          Object {
            "location": Object {
              "end": Object {
                "column": 27,
                "line": 4,
              },
              "start": Object {
                "column": 23,
                "line": 4,
              },
            },
            "type": "identifier",
            "value": "bazz",
          },
          Object {
            "location": Object {
              "end": Object {
                "column": 33,
                "line": 4,
              },
              "start": Object {
                "column": 29,
                "line": 4,
              },
            },
            "type": "identifier",
            "value": "buzz",
          },
        ],
        "type": "list",
      },
      "type": "assignment",
    },
    Object {
      "lhs": Object {
        "location": Object {
          "end": Object {
            "column": 14,
            "line": 5,
          },
          "start": Object {
            "column": 11,
            "line": 5,
          },
        },
        "type": "identifier",
        "value": "bar",
      },
      "rhs": Object {
        "location": Object {
          "end": Object {
            "column": 20,
            "line": 5,
          },
          "start": Object {
            "column": 17,
            "line": 5,
          },
        },
        "type": "identifier",
        "value": "car",
      },
      "type": "assignment",
    },
    Object {
      "lhs": Object {
        "location": Object {
          "end": Object {
            "column": 14,
            "line": 6,
          },
          "start": Object {
            "column": 11,
            "line": 6,
          },
        },
        "type": "identifier",
        "value": "jar",
      },
      "rhs": Object {
        "location": Object {
          "end": Object {
            "column": 23,
            "line": 6,
          },
          "start": Object {
            "column": 17,
            "line": 6,
          },
        },
        "type": "string",
        "value": "test",
      },
      "type": "assignment",
    },
  ],
  "type": "MagentoDirective",
}
`;
